test/ruby/test_exception.rb

```diff
@@ -924,8 +924,14 @@ def capture_warning_warn
       alias_method :warn2, :warn
       remove_method :warn
 
-      define_method(:warn) do |str|
-        warning << str
+      if category
+        define_method(:warn) do |str, category:|
```

Comment:
Can we use `**kw` and `kw[:category]` to reflect the default behavior of `Warning.warn`?  Alternatively, make `Warning.warn` only accept the `:category` keyword, and use `category: nil`.